Initial specifications for socks forward mode

Anisotrop opened the pull request.

Commit:
First shot at doing an initial implementation, 1 control script and 1 connection established script partially complete

Anisotrop committed:
Fix TODO in stream test.

StCostea committed:
Socks nukleus - tests for control.

Anisotrop committed:
Implement feedback.

Commit:
Merge pull request #2 from Anisotrop/dpwspoon.feature/initial-impl

Fix TODO in stream test.

Commit:
Merge pull request #1 from StCostea/socks

Socks nukleus - tests for control.

Anisotrop committed:
Streams test was not accepting correctly in implementation.

Anisotrop committed:
Send negotition request fragmented.

Anisotrop committed:
Add destination address and port as extension on the route. Update and clean tests.

Anisotrop committed:
Fix of tests with respect of 'external scope' logic.

Anisotrop committed:
Add destAddrPort as extension on all controller scripts.

Anisotrop committed:
Added more data to streams test.

Anisotrop committed:
Add tcp nukleus as dependency. Refactoring for socks forward mode.

Anisotrop committed:
Use TCP Nukleus for SOCKS server forward mode.

dpwspoon commented on pom.xml:
```diff
@@ -11,6 +11,9 @@
   <description>SOCKS Nukleus Specification</description>
   <url>https://github.com/reaktivity/nukleus-socks.spec</url>
 
+<inceptionYear>2017
```
please fix formatting

dpwspoon commented on src/main/scripts/org/reaktivity/specification/socks/rfc1928/forward/client.connect.send.data/server.rpt:
```diff
@@ -0,0 +1,60 @@
+write [0x00] # RSV field
+write [0x01] # Address Type domain name
+write [0x7F 0x00 0x00 0x01] # The Bound Server Address
+write [0xEA 0xDB]           # The Bound Server Port
```
Isn't this suppose to be the ephemeral port?  How do we get this?

Anisotrop replied:
On the spec project it is sent back by `server.rpt` in the same directory. 

In the implementation test it will be taken from `.../specification/nukleus/socks/streams/forward/client.connect.send.data/server.rpt`. This is because the socks nukleus, acting as `server`, will use the `TcpBeginExFW` of the response from target to construct the Socks Connect Response. 

When acting as `client`, it receives this information as an already encoded socks message but makes no use of it (as already discussed).

Please see below a diagram of the `server` and `client` behaviors in `forward` mode. The example routing might need more precision, but it should not contain errors affecting the socks spec.

dpwspoon commented on src/main/scripts/org/reaktivity/specification/nukleus/socks/streams/forward/client.connect.send.data/client.rpt:
```diff
@@ -0,0 +1,35 @@
+  option nukleus:window 65536
+  option nukleus:transmission "duplex"
+
+# Client forward mode DOES NOT enforce target Nukleus to be TCP
```
I think this is false for SOCKS RFC client.  The target needs to be TCP or UDP (no UDP nukelus yet).  The response will have the tcp extensions with the local and remote addresses.  I.E. read nukleus ext.  This is needed for the ephemeral port on the SOCKS CONNECT in https://github.com/reaktivity/nukleus-socks.spec/pull/1/files#diff-de9f689af65764d57451a65b210fe493 right?

Anisotrop replied:
The BeginFrame of the target, coming on the connect-reply stream is only used to initiate the socks negotiation and connect steps. The connect response also has the bind address that the remote socks server has allocated, but for now it is not specified how to use.

dpwspoon commented on src/main/resources/META-INF/reaktivity/socks.idl:
```diff
@@ -0,0 +1,32 @@
+        struct SocksRouteEx
+        {
+            socks::SocksMode mode;
+            string16 destAddrPort;
```
Is this needed?  And what for?

If it is needed, I would suggest using the same address ext  (maybe the same extension) as TCP: 
```
            tcp::TcpAddress localAddress;
            uint16 localPort;
```

https://github.com/reaktivity/nukleus-tcp.spec/blob/develop/src/main/resources/META-INF/reaktivity/tcp.idl

Anisotrop replied:
This is an attempt to match the 3 fields in the [Socks RFC](https://tools.ietf.org/html/rfc1928#section-4) since in case of `0x03` for `ATYP` a domain name should be sent. The initial idea was to describe the whole 3 fields (`ATYP`, `DST.ADDR`, `DST.PORT`) but went for a simpler String mapping.

dpwspoon replied:
Ok with me.  Thoughts @jfallows

dpwspoon commented:
Please resolve merge conflicts.

Anisotrop committed:
Fix formatting.

Anisotrop committed:
Fixed conflicts.

Anisotrop commented:
Below is a proposed behavior of the sokcs nuklei as server and client in forward mode. specifying only the `CONNECT` part of the [RFC](https://tools.ietf.org/html/rfc1928), while `BIND` is not covered yet.

The routes are built for 1 source TCP nukeuls, a Socks Nukleus and 2 target TCP nukeli.  The idea would be to have the possibility to connect to `example.com` or `example.net` via the route on either client or server. 

```
    On TCP#1
    ROUTE0 (SOURCE=?, TARGET=SOCKS#1, TcpAddressExt=?)

    On SOCKS#1:
    ROUTE1  (SERVER, SOURCE=TCP#1, TARGET=TCP#2, Mode=Forward, DstAddrPort=example.com:8080)
    ROUTE3  (CLIENT, SOURCE=TCP#1, TARGET=TCP#3, Mode=Forward, DstAddrPort=example.net:8080)
    
    On TCP#2:
    ROUTE20 (SOURCE=SOCKS#1, TARGET=?, TcpAddressExt=nslookup(example.com):8080)
    
    On TCP#3:
    ROUTE30 (SOURCE=SOCKS#1, TARGET=?, TcpAddressExt=nslookup(example.net):8080)
```
    
Some notes:
- It is not clear to me if TCP#1 should in fact consist of 2 nuklei.
- The target TCP#3 nukleus could be any Nukleus that returns a begin frame only after it's own connection to the target has been established.


Socks server
```
ROUTE0, ROUTE1, ROUTE20                                            
SOURCE(TCP#1)                              SOCKS(SOCKS#1)                         TARGET(TCP#2)
                         (read)                               (write)            
     CONNECT -----------------> ACCEPT          ||    CONNECT  -----------------> ACCEPT
                1. BF                           ||        6. BF
                3. DF (SOCKS NEGOTIATE REQUEST) ||    
                5. DF (SOCKS CONNECT REQUEST)   ||          
                                                ||          
                         (write)                ||                   (read)
     CONNECT-REPLY <----------- ACCEPT-REPLY    ||    CONNECT-REPLY  <---------- ACCEPT-REPLY
                2. BF                           ||        7. BF (TcpBeginExFW)
                4. DF (SOCKS NEGOTIATE RESPONSE)||    
                                                ||          
                8. DF (SOCKS CONNECT RESPONSE)  ||    
```              
   
Socks client
```
ROUTE0, ROUTE3, ROUTE30                                            
SOURCE(TCP#1)                              SOCKS(SOCKS#1)                         TARGET(TCP#3)
                         (read)                               (write)            
     CONNECT -----------------> ACCEPT          ||    CONNECT  -----------------> ACCEPT
                1. BF                           ||        2. BF
                                                ||        4. DF (SOCKS NEGOTIATE REQUEST)
                                                ||        6. DF (SOCKS CONNECT REQUEST)  
                                                ||          
                         (write)                ||                   (read)
     CONNECT-REPLY <----------- ACCEPT-REPLY    ||    CONNECT-REPLY  <---------- ACCEPT-REPLY
                8. BF                           ||        3. BF (TcpBeginExFW)
                                                ||        5. DF (SOCKS NEGOTIATE RESPONSE)    
                                                ||        7. DF (SOCKS CONNECT RESPONSE) 
```

dpwspoon commented:
> It is not clear to me if TCP#1 should, in fact, consist of 2 nuklei.

In general, there is only on TCP nukleus with many different routes.  So TCP#0 and TCP#2 and TCP#3 will all be going through the same nukelus instance but have different routes.  

Routes have source, sourceRef, target and targetRef.  In your example, you have TCP#0 going to Route1 and Route2 of SOCKS#1.  If I'm interpreting your question correctly, you are asking can you have TCP#0 (same source, sourceref, target and targetRef)  where target="Socks" and targetRef="SomeSocksRef" be routed out of socks in two different paths.  I would think **NO**, SOCKS should not be routed like this because there are no indicators inherent on the begin frame for whether it wants server or client mode or where to route it.  

No nuklei today does routing in the way you implied where it could be a client one way and a server the other.  In contrast, HTTP Begin frames have headers, so http can route one route in two multiple routes out for either client OR server, but not both!  

> The target TCP#3 nukleus could be any Nukleus that returns a begin frame only after it's own connection to the target has been established.

Makes sense to me that TCP is not required.  Thanks for explaining.

It also looks like to me TCP#1 does not need to be TCP.

Anisotrop committed:
Fix wrongly merged conflicts with upstream.

Anisotrop commented:
Hello @dpwspoon, yes, TCP#1 in the diagram above is not required to be TCP.

I have updated the client diagram to take throttling into account.

```
SOURCE(TCP#1)                              SOCKS(SOCKS#1)                         TARGET(TCP#3)
                                                ||
                         (read)                 ||            (write)            
     CONNECT -----------------> ACCEPT          ||    CONNECT  -----------------> ACCEPT
                1. > BF >-----------------------||------> 1. > BF
                2. < WF <--------------------+  ||        2. > DF (WF*) (SOCKS NEGOTIATE REQUEST)
                                             |  ||  +---> 3. > DF (WF*) (SOCKS CONNECT REQUEST)      
                                             |  ||  |     
                3./4. > DF >-----------------|--||--|---> 4./5. > DF
                3./4. < WF <-----------------|--||--|---< 4./5. < WF  
                                             |  ||  |
                                             |  ||  |
                         (write)             |  ||  |                (read)
     CONNECT-REPLY <----------- ACCEPT-REPLY |  ||  | CONNECT-REPLY  <---------- ACCEPT-REPLY
                1. BF <----------------------|--|| -|---< 1. < BF (TcpBeginExFW)
                                             |  ||  +---< 2. < DF (SOCKS NEGOTIATE RESPONSE)    
                                             +--||------< 3. < DF (SOCKS CONNECT RESPONSE) 
                                                ||    
                2./3. < DF <--------------------||------< 4./5. < DF
                2./3. > WF >--------------------||------> 4./5. > WF
                                                ||
                                                
    WF* - enough WINDOW frames are received to allow sending the DATA frames on the stream                                            
          if it is not the case, the DATA frame is sent after a new WINDOW is received and
          the condition is met
                                                
```

dpwspoon commented on src/main/resources/META-INF/reaktivity/socks.idl:
```diff
@@ -0,0 +1,32 @@
+ */
+scope socks
+{
+    enum SocksMode
```
I don't think Mode will be needed.  As this PR only implements FORWARD tests, I would suggest removing Mode for now.  And adding it back if in the unlikely event it is needed in the future.

dpwspoon commented:
If we are not using TcpExtension based on https://github.com/reaktivity/nukleus-socks.spec/pull/1#discussion_r141915168, which I agree is a good idea, then we should no longer have a dependency on TcpExtension, and should migrate all to SocksExtensions.

Following the conversation of the HttpBeginFrames.  We should follow the model of the other nuklei, and add the extension meta-data SocksAddressEX to the stream BEGIN frames and the Route frames.  By doing this, the SocksAddressEX becomes optional on the Route and the BEGIN frames coming into the SocksClientAccept and the SocksServerConnectReply. 

![ipad-socks-forward-drawing](https://user-images.githubusercontent.com/2014623/31454766-69635080-ae6b-11e7-9234-85830fd1df91.png)

For the client, the address extension on the SocksRoute and the SocksBeginEX goes on the SOCKS Connect frame.  The address that comes back on the SOCKS connected goes out on the SocksBeginEx on the ClientAcceptReply.

For the server, if there is an address on the route command then it must match the address on the Socks CONNECT frame.  It the address is not specified on the route, then it is not enforced and it is forwarded.  The Address would be put on the SocksBeginEx of the SocksServerConnect.

If there is a SocksBeginEX on the ServerConnectReply, then this is what should be put on the SOCKS CONNECTED frame.  If there is not then we can put something arbitrary (maybe the address that came in)

This address on the SOCKS Connected frame should be added to the SocksBeginEX for the ClientAcceptReply.

(apologies for my drawing, did it on the ipad late at night)

By doing this model, we will allow the SOCKS server to have many different accepts from the same source and target, but routed to different connects based on address.  This should be "best match", so all addresses with the extensions are compared first, and then there can be a catch-all route without the extension (This is what http does as well).  

Reading the SocksBeginEX on the ServerConnectReply might be problematic as this will most likely be a TcpBeginEX.  As such, I would consider that part of the implementation blocked (to be completed elsewhere).

I also think the adding of the address to the Route is the general case, we can support for doing on the Begins in a follow up PR if you want to limit the scope of this PR.

dpwspoon commented:
I've filed the story/question for transforming the metadata https://github.com/reaktivity/ignition/issues/8.

dpwspoon reviewed (changes requested):
Blocking changes requested: 
- [ ] Please remove mode in the socks.idl.
- [ ] Please add the Socks Extension to the Begin Frame.  
- [ ] Please remove dependency on TcpEX

Anisotrop committed:
Upgrade K3PO.

NicoletaOita committed:
Added tests for socks nukleus

Anisotrop committed:
Upgrade dependencies and fix scripts.

Anisotrop committed:
Build with Java9.

Anisotrop committed:
Merge pull request #1 from NicoletaOita/socks_nukleus_tests

Added tests for socks nukleus

Anisotrop committed:
Add low-level test ann explicit maven-compiler-plugin version.

Anisotrop committed:
Merge branch 'impl.continued' of https://github.com/Anisotrop/nukleus-socks.spec into impl.continued

Anisotrop committed:
Small fixes in scripts and added them in client also.

Anisotrop committed:
Remove 'mode' from route.

Anisotrop committed:
Tests with different window size for throttling support.

Anisotrop committed:
Fix test to send unsupported method.

Anisotrop committed:
Enhance throttling tests to have different window sizes for both server and client.

Anisotrop committed:
Added throttling tests.

Anisotrop committed:
Update flow control to credit based.

Anisotrop committed:
More tests for throttling.

dpwspoon commented on pom.xml:
```diff
@@ -44,11 +45,77 @@
+    <dependency>
+      <groupId>javax.annotation</groupId>
+      <artifactId>javax.annotation-api</artifactId>
+      <version>1.3.1</version>
```
nit: use property for version to match others?

dpwspoon commented on src/main/scripts/org/reaktivity/specification/nukleus/socks/streams/forward/client.connect.send.data.throttling.client.smaller/client.rpt:
```diff
@@ -0,0 +1,47 @@
+read "data flows back as intended"
+
+## 60000 bytes Application data
+write "LongText110000000010LongText110000000010LongText110000000010LongText110000000010LongText110000000010LongText110000000010LongText110000000010LongText110000000010LongText110000000010LongText110000000010LongText110000000010LongText110000000010LongText110000000010LongText110000000010LongText110000000010LongText110000000010LongText110000000010LongText110000000010LongText110000000010LongText1100 ...
```
suggest writing a simple function to write this data, https://github.com/reaktivity/nukleus-http-cache.spec/blob/develop/src/main/java/org/reaktivity/specification/nukleus/http_cache/specification/internal/Functions.java#L113

dpwspoon commented on src/main/scripts/org/reaktivity/specification/nukleus/socks/streams/forward/client.connect.send.data.throttling.client.smaller/server.rpt:
```diff
@@ -0,0 +1,58 @@
+write  "data flows back as intended"
+
+## 60000 bytes Application data
+read "LongText110000000010LongText110000000010LongText110000000010LongText110000000010LongText110000000010LongText110000000010LongText110000000010LongText110000000010LongText110000000010LongText110000000010LongText110000000010LongText110000000010LongText110000000010LongText110000000010LongText110000000010LongText110000000010LongText110000000010LongText110000000010LongText110000000010LongText11000 ...
```
Same for all

dpwspoon commented on src/test/java/org/reaktivity/specification/nukleus/socks/stream/forward/ConnectionIT.java:
```diff
@@ -0,0 +1,135 @@
+
+    @Test
+    @ScriptProperty("serverAccept \"nukleus://socks/streams/source\"")
+    @Specification({
```
consider pull out the spec RFC cases into a different class then the nukelus behavior tests

cmebarrow commented on src/main/java/org/reaktivity/specification/socks/internal/Functions.java:
```diff
@@ -0,0 +1,126 @@
+{
+
+    @Function
+    public static byte[] destAddrPort(String destAddrPort)
```
If I understand correctly, all this function is doing is converting a string (any string) into a string16. So I would suggest naming it accordingly, something like asString16. But see also my comment on socks.idl, which will likely change the requirements for this function, I am actually not sure we should have this function at all. Unless it's very clumsy it's usually better to have the scripts write bytes directly rather than calling functions, since it makes it easier to know exactly what bytes are going on the wire. Follow the pattern of other existing scripts, for example those in nukleus-tcp.spec.

cmebarrow commented on src/main/resources/META-INF/reaktivity/socks.idl:
```diff
@@ -0,0 +1,25 @@
+    {
+        struct SocksRouteEx
+        {
+            string16 destAddrPort;
```
For consistency with the tcp nukleus, this should be a union, similar to the one in [tcp.idl](https://github.com/reaktivity/nukleus-tcp.spec/blob/develop/src/main/resources/META-INF/reaktivity/tcp.idl), to deal explicitly with the 3 possible address formats as described in https://www.ietf.org/rfc/rfc1928.txt. Just copy the union from tcp then add the 3rd member, which should be a regular string, not string16, since SOCKS address length is specified as a single byte so cannot exceed that.

Same would apply later when we include this in SocksBeginEx.

Anisotrop committed:
Implemented code review feedback.

cmebarrow commented on src/main/java/org/reaktivity/specification/socks/internal/Functions.java:
```diff
@@ -0,0 +1,151 @@
+
+    /**
+     * Builds the TcpBeginExFW from given IPv4 addresses
+     * FIXME Should be implemented in tcp.spec nukleus, with tests on both client and server.
```
The TCP begin extension is currently only used to inform users on the TCP nukleus what the local and remote addresses and ports are, for an incoming network connection (server accept stream), when tcp nukleus is acting as a server, and for the reply to an outgoing network connection (client connect reply stream) when the TCP nukleus is acting as a client. Users of the TCP nukleus do not currently ever set this data, and if they do it has no effect.him So you should not need this function for scripts testing the socks nukleus acting as a client.

I do understand that you might need it for scripts testing the socks nukleus acting as a server, until such time as we use SocksBeginEx instead.

Anisotrop replied:
Yes, this is needed so that the socks server would send back to the client the address and port it has used to connect to the requested address.

> CONNECT
> 
>    In the reply to a CONNECT, BND.PORT contains the port number that the
>    server assigned to connect to the target host, while BND.ADDR
>    contains the associated IP address.  The supplied BND.ADDR is often
>    different from the IP address that the client uses to reach the SOCKS
>    server, since such servers are often multi-homed.  It is expected
>    that the SOCKS server will use DST.ADDR and DST.PORT, and the
>    client-side source address and port in evaluating the CONNECT
>    request.

cmebarrow commented on src/main/scripts/org/reaktivity/specification/nukleus/socks/control/route/client/controller.rpt:
```diff
@@ -0,0 +1,49 @@
+write [0x06] "target"
+write ${newClientConnectRef}
+write 0L # authorization
+write ${socks:destAddrPort(destAddrPort)}
```
This can just be 
```
write [0x0b] "example.com"
```
you don't need the function
(actually once you put in the union that would be the extra byte for type enum, so 
`write [0x03 0x0b] "example.com"`
Same applies in the other scripts

cmebarrow commented on src/main/java/org/reaktivity/specification/socks/internal/Functions.java:
```diff
@@ -0,0 +1,151 @@
+     * @return
+     */
+    @Function
+    public static byte[] beginTcpExtensionIpv6(
```
You should not need this, see my comment on the other function

cmebarrow commented on src/main/scripts/org/reaktivity/specification/nukleus/socks/streams/forward/client.connect.send.data.throttling.client.smaller/client.rpt:
```diff
@@ -0,0 +1,47 @@
+connect await ROUTED_CLIENT
+        "nukleus://socks/streams/source"
+  option nukleus:route ${newClientAcceptRef}
+  option nukleus:window 40000
```
Please use a much smaller value. Then you will not need to generate long data frames. Should be able to force the same code paths that way, and the test scripts will be much easy to understand and debug. Basically each of our tests should be as simple as possible, no more complicated than they need to be to test what we are testing. The mathematics of Windows and flow control are independent of the actual size of the numbers.

cmebarrow commented on src/main/scripts/org/reaktivity/specification/nukleus/socks/streams/forward/client.connect.send.data.throttling.client.smaller/client.rpt:
```diff
@@ -0,0 +1,47 @@
+read "data flows back as intended"
+
+## 60000 bytes Application data
+write ${socks:generateLongText(60000)}
```
See my comment above, there is no value in generating huge values.

cmebarrow commented on src/main/scripts/org/reaktivity/specification/nukleus/socks/streams/forward/client.connect.request.with.command.not.supported/server.rpt:
```diff
@@ -0,0 +1,38 @@
+accepted
+
+# Server forward mode DOES enforce target Nukleus to be TCP by sending the TcpBeginExFW
+write nukleus:begin.ext ${socks:beginTcpExtensionIpv4(serverLocalAddress, serverLocalPort, serverRemoteAddress, serverRemotePort)}
```
I don't understand this. Is the purpose to report the address and port that were used to satisfy the CONNECT request?

Anisotrop replied:
Yes, this is needed so that the socks server would send back to the client the address and port it has used to connect to the requested address.

> CONNECT
> 
>    In the reply to a CONNECT, BND.PORT contains the port number that the
>    server assigned to connect to the target host, while BND.ADDR
>    contains the associated IP address.  The supplied BND.ADDR is often
>    different from the IP address that the client uses to reach the SOCKS
>    server, since such servers are often multi-homed.  It is expected
>    that the SOCKS server will use DST.ADDR and DST.PORT, and the
>    client-side source address and port in evaluating the CONNECT
>    request.

cmebarrow replied:
@Anisotrop David says we can use the address and port from the route (SocksRouteEx) instead. That way we can remove the dependency on tcp. Eventually this will be SocksBeginEx but we don't need to do that yet. We just need to wrap up this socks forward work so you can get onto the socks reverse stuff, which is our most pressing need (for kaazing.net). So please remove this line (as also commented elsewhere).

cmebarrow commented:
@Anisotrop @dpwspoon @jfallows 
I am reviewing Costin's PRs for socks in forward mode based on the following assumptions. If these assumptions are not correct please let me know.

* socks as client (no buffering required): 
  *  accept side can be any nukleus stream (for now, later we will support a socks specific begin extension to dynamically connect to any address and port)
  *  connect side talks SOCKS protocol (over any transport)
  *  routing: routeEx contains the DST.ADDR and DST.PORT that should be written in the SOCKS Connect request on the client connect stream (later this will be optional when we can get that information from the SocksBeginEx)
  *  client accept: Begin processing looks up the route, starts target (client connect) stream, waits for window, then writes SOCKS CONNECT request. Waits for response (on client connect reply). If successful, sends Window (whatever remains from the client connect stream), else Reset.
  *  client connect reply: does nothing on begin. On receiving SOCKS CONNECT reply, tells accept stream to send window, and sends begin on client accept reply.
* socks as server:
  *  purpose is to act as a socks server
  *  accept side talks SOCKS protocol (over any nukleus transport)
  *  connect side talks to tcp nukleus (later it will surface BND.ADDR and BND.PORT from the incoming SOCKS Connect request into a high level socksBeginEx; this will get transformed by a new concept of transformer into TcpBeginEx metadata if routed into tcp)
  *  We could support dynamic routing, but based on the implementation so far and [this comment](https://github.com/reaktivity/nukleus-socks.spec/pull/1#issuecomment-335886861) I will assume we don't, so a route has to exist matching the DST.ADDR and DST.PORT of the incoming socks CONNECT request.
  *  server accept: begin processing sends a window sufficient to receive the largest possible SOCKS Connect request. Must buffer incoming data in case of (a) incomplete connect request (b) additional data follows the Socks connect request. On receiving SOCKS CONNECT request, looks for an available route specifying the same DST.ADDR and DST.PORT. If found, starts the target (server connect) stream by writing a TCP-style BEGIN frame with target = (address or host name from socks Connect or the route), targetRef = (port from socks Connect/route). On receiving a window, writes any data we already received (after the connect request).
  *  server connect reply: on begin start the correlated server accept reply stream. No buffering or flow control needed, all streams and throttle frames are just forwarded as is.

cmebarrow commented on src/main/scripts/org/reaktivity/specification/socks/rfc1928/forward/client.connect.send.data.throttling.window.8.padding.7/client.rpt:
```diff
@@ -0,0 +1,139 @@
+read [0x7F 0x00 0x00 0x01] # The Bound Server Address
+read [0xEA 0xDB]           # The Bound Server Port
+
+## Application data
```
I get a bit worried when I see scripts throwing in a lot of data around like this, it makes me wonder if there's a clear understanding of the purpose of the test. It should be possible to force the code paths you want to force without sending dozens of data frames. Why will sending 30 "hellworld" 's really reveal a bug that sending 3 or 5 or 10 would not? If so, why not try with 300 or a million? Our spec tests should be targeting specific known edge cases, not trying to throw the kitchen sink at it and hoping to find something wrong. That's the job of stress tests or performance tests.

Anisotrop replied:
The idea would be to have the conditions requiring to start buffering (based on window/credit options) and an initial handshake window size. Once the sender's window is exhausted, buffering is no longer necessary and frames can be forwarded. The amount of data in these tests is 400 bytes which for an initial window of the same size or less should trigger the above cases. Also, in real case, a message could be 261 bytes so an initial window less than this also seems too small.

cmebarrow reviewed (changes requested):
This is comments I have to far. Please see also my general comment on the PR, which is to check my understanding.

Anisotrop commented:
Hello @cmebarrow, thanks for the review. The client doesn't need buffering if the Window received on the Accept-Reply is larger than the individual responses involved in the handshake. If it is smaller and forwards it, the handshake will stall. To prevent this, the window should not rely on the accept reply but send it's own initial swize. If it is larger than the accept-reply's window buffering should occur. Let me try to illustrate:

```
CLIENT (WITH THROTTLING)
SOURCE(NUKLEUS#1)                         SOCKS(SOCKS#1)                         TARGET(TCP#3)
                                                ||
                                                
                          [Accept/Connect exchanges omitted for readability]
                                  
                         (write)             |  ||  |                (read)
     CONNECT-REPLY <----------- ACCEPT-REPLY |  ||  | CONNECT-REPLY  <---------- ACCEPT-REPLY
                1. BF <----------------------|--|| -|---< 1. < BF (TcpBeginExFW)
                
                
                
            CASE A: Initial window larger than any socks handshake message
                WF (100) ----------------- Forward initial window -----------------------> WF (100)
                                                                             (Available window 100)
                                             |  ||  +---< 2. < DF (2)(SOCKS NEGOTIATE RESPONSE)
                                             ----------------------------------------------> WF (2)   
                                                                             (Available window 100)
                                             +--||------< 3. < DF (20) (SOCKS CONNECT RESPONSE) 
                                             ---------------------------------------------> WF (20)
                                                                             (Available window 100)
                Available window (100)                             
                Handshake is done
            END of CASE A
                                             
                                             
            CASE B: Initial window smaller than a socks handshake message
                WF (5) ------------------- Forward initial window -----------------------> WF (5)
                                                                             (Available window 5)
                                             |  ||  +-----< 2. < DF (2)(SOCKS NEGOTIATE RESPONSE)
                                             --------------------------------------------> WF (2)
                                                                             (Available window 5)                                             
                                             STALLED -----< 3. < DF (20) (SOCKS CONNECT RESPONSE)
                                                                             (Available window 5)
                Handshake has failed
            END of CASE B

            CASE B': Initial window smaller than a socks handshake message
                WF (5) -------------------     Send initial window -----------------------> WF (500)
                                                                             (Available window 500)
                                             |  ||  +-----< 2. < DF (2)(SOCKS NEGOTIATE RESPONSE)
                                             --------------------------------------------> WF (2)
                                                                             (Available window 500)                                             
                                             +--||------< 3. < DF (20) (SOCKS CONNECT RESPONSE)
                                             ---------------------------------------------> WF (20)
                                                                             (Available window 500)
                Handshake is done, should buffer 495 bytes.
            END of CASE B
```

Anisotrop committed:
Changed SocksRouteEx to match SOCKS protocol definition.

cmebarrow commented on src/main/resources/META-INF/reaktivity/socks.idl:
```diff
@@ -0,0 +1,33 @@
+    {
+        case 1: octets[4] ipv4Address;
+        case 3: string domainName;
+        case 4: octets[16] ipv6Address;
```
Why the gap in numbering? I would recommend 
```
        case 1: octets[4] ipv4Address;
        case 2: octets[16] ipv6Address;
        case 3: string domainName;
```
so it's like tcp only with just case 3 added.

Anisotrop replied:
It is done like this to match the `ATYP` field specified in https://tools.ietf.org/html/rfc1928#section-4.

cmebarrow replied:
I see, but I'm not sure if there is any value in that. @jfallows WDYT?

cmebarrow reviewed (changes requested):
I having gone through it all yet. This is what I have so far. Your test failures in ControllerIT (in the related PR) are due to a bug in the flyweight generator (nukleus-maven-plugin). The generated code for SocksAddressFW.Builder.domainName(string value) is incorrectly calling the domainName() method multiple times. I was able to ffix this temporarily by changing the generated method to the following:
```
    public Builder domainName(String value) {
      if (value == null) {
        limit(offset() + FIELD_OFFSET_DOMAIN_NAME);
      } else {
        kind(KIND_DOMAIN_NAME);
        StringFW.Builder domainName = domainName();
        domainName.set(value, StandardCharsets.UTF_8);
        limit(domainName.build().limit());
      }
      return this;
    }
```

Anisotrop committed:
Remove no longer needed tests.

Anisotrop committed:
Remove unused functions.

cmebarrow commented on src/main/scripts/org/reaktivity/specification/nukleus/socks/control/route/client/controller.rpt:
```diff
@@ -0,0 +1,48 @@
+write ${newClientConnectRef}
+write 0L # authorization
+write [0x03 0x0b] "example.com"
+write [0x1F 0x90]
```
k3po and k3po nukleus transport now support the use of explicit syntax to write a short (two byte integer) in native byte order, matching the type in the IDL (which is uint16), so you should one of the supported forms
```
write 8080s
write short 8080
```
The change made to k3po to support this is https://github.com/k3po/k3po/pull/455. I have also now documented it in https://github.com/k3po/k3po/wiki/Scripting-Language.

cmebarrow commented on src/main/scripts/org/reaktivity/specification/nukleus/socks/control/route/client/nukleus.rpt:
```diff
@@ -0,0 +1,44 @@
+read (long:clientConnectRefRG010101N)
+read [0..8] # authorization
+read [0x03 0x0b] "example.com"
+read [0x1F 0x90]
```
read short 8080

cmebarrow commented on pom.xml:
```diff
@@ -44,11 +45,79 @@
     <jacoco.coverage.ratio>1.00</jacoco.coverage.ratio>
     <jacoco.missed.count>0</jacoco.missed.count>
+    <junit.version>4.12</junit.version>
+    <k3po.version>3.0.0-alpha-84</k3po.version>
```
Use 3.0.0-alpha-85 or later to get the new syntax for reading and writing shorts (etc)

cmebarrow commented on pom.xml:
```diff
@@ -111,6 +184,11 @@
+          <excludes>
+            <exclude>org/reaktivity/specification/socks/internal/types/**/*</exclude>
+            <exclude>org/reaktivity/specification/socks/internal/Functions.*</exclude>
+            <exclude>org/reaktivity/specification/socks/internal/Functions$Mapper.*</exclude>
```
John doesn't want us to add these excludes, because we are planning to generate unit tests for the flyweights eventually.

cmebarrow commented on pom.xml:
```diff
@@ -44,11 +45,79 @@
+    <nukleus.spec.version>0.11</nukleus.spec.version>
+    <nukleus.plugin.version>0.13</nukleus.plugin.version>
+    <nukleus.tcp.spec.version>0.21</nukleus.tcp.spec.version>
+    <maven-enforcer-plugin.version>1.3.1</maven-enforcer-plugin.version>
```
We are using version 3.0.0-M1 in http.spec and tls.spec

cmebarrow commented on src/main/scripts/org/reaktivity/specification/nukleus/socks/control/route/server/controller.rpt:
```diff
@@ -0,0 +1,50 @@
+
+property controlRG010001C ${nukleiRG010001C.control("socks")}
+
+property destAddrPort "example.com:8080"
```
This is not used

cmebarrow commented on src/main/scripts/org/reaktivity/specification/nukleus/socks/control/route/server/controller.rpt:
```diff
@@ -0,0 +1,50 @@
+write ${newServerConnectRef}
+write 0L # authorization
+write [0x03 0x0b] "example.com"
+write [0x1F 0x90]
```
write short 8080 or write 8080s (please apply this everywhere)

cmebarrow commented on src/main/scripts/org/reaktivity/specification/nukleus/socks/control/unroute/client/controller.rpt:
```diff
@@ -0,0 +1,47 @@
+write [0x06] "target"
+write ${newClientConnectRef}
+write 0L
+write [0x03 0x0b] "example.com"
```
We'll need variants of this script (and the other scripts) to cover the 3 address types

cmebarrow commented on src/main/scripts/org/reaktivity/specification/nukleus/socks/streams/forward/client.connect.fallback.to.no.authentication/server.rpt:
```diff
@@ -0,0 +1,46 @@
+accepted
+
+# Server forward mode DOES enforce target Nukleus to be TCP by sending the TcpBeginExFW
+write nukleus:begin.ext ${socks:beginTcpExtensionIpv4(serverLocalAddress, serverLocalPort, serverRemoteAddress, serverRemotePort)}
```
I see, you need the connected to address and port to encode the SOCKS connect reply on the wire. Hence the need for a dependency on tcp.spec. I will discuss this tomorrow with John to see if we can avoid the dependency.

cmebarrow replied:
Discussed with him and David. Instead socks nukleus just usethe address and port from the route (socks route extension) when encoding the reply. So you can remove this line altogether. Same applies to other similar scripts.

cmebarrow reviewed (changes requested):
This is what I have so far. I'll need to discuss with John to finish reviewing.

Anisotrop committed:
Add more test cases for (domain, ipv4, ipv6), upgrade k3po.

cmebarrow commented on src/main/scripts/org/reaktivity/specification/nukleus/socks/streams/forward/client.connect.fallback.to.no.authentication/client.rpt:
```diff
@@ -0,0 +1,35 @@
+  option nukleus:window 65536
+  option nukleus:transmission "duplex"
+
+# Client forward mode DOES NOT enforce target Nukleus to be TCP
```
Please remove this comment here and similar ones elsewhere

cmebarrow commented on src/main/scripts/org/reaktivity/specification/nukleus/socks/streams/forward/client.connect.fallback.to.no.authentication/server.rpt:
```diff
@@ -0,0 +1,46 @@
+
+property serverAccept "nukleus://target/streams/socks#source"
+
+# Build the TcpBeginExFW using hardcoded values
```
Just use the host and port from the route instead when encoding the connect reply (like address "example.com" port 8080). We don't need full blown socks vs tcp support at this time (Later when we do want to add that we will use SocksBeginEx plus a transformer or fuser to transform the extension metadata according to the target nukleus).

cmebarrow commented on src/main/java/org/reaktivity/specification/socks/internal/Functions.java:
```diff
@@ -0,0 +1,78 @@
+/**
```
This class will no longer be needed once you remove all writing of tcp begin ex.

cmebarrow commented on pom.xml:
```diff
@@ -42,13 +43,85 @@
+    </dependency>
+    <dependency>
+      <groupId>org.reaktivity</groupId>
+      <artifactId>nukleus-tcp.spec</artifactId>
```
Please remove this dependency. See my comments on scripts using tcpBeginEx for how we can avoid the need for it.

cmebarrow reviewed (changes requested):
@Anisotrop All of my comments revolve around removing the dependency on tcp nukleus (spec). Ping me on slack tomorrow if you have any questions or need to talk.

Anisotrop committed:
Remove tcp nukleus dependency and k3po read/write short syntax.

Anisotrop commented:
Hello @cmebarrow, I have removed the tcp nukleus dependency. Also, I have reverted the usage of the read/write `short` in the scripts. From what I have seen they are producing values in `little-endian` but the port is specified in the RFC as `big-endian`. Also, for now, I have set the bound address and port as 0. I think the scripts were becoming confusing with an address repeated that many times. Since we cannot use the correct info, I thought we could use a simpler approach.

cmebarrow commented:
@Anisotrop You can use the new k3po-nukleus-ext option nukleus:byteorder to have integer values be written and read in network byte order (big-endian). This was recently added, see https://github.com/reaktivity/k3po-nukleus-ext.java/pull/29.

cmebarrow commented on src/test/java/org/reaktivity/specification/nukleus/socks/control/ControlIT.java:
```diff
@@ -0,0 +1,169 @@
+        k3po.finish();
+    }
+
+    @Test
```
Need to add a negative test for unroute where no matching route is found, along the lines of https://github.com/reaktivity/nukleus-auth-jwt.spec/tree/develop/src/main/scripts/org/reaktivity/specification/nukleus/auth/jwt/control/unroute/proxy/fails.unknown.route

cmebarrow commented on src/main/scripts/org/reaktivity/specification/nukleus/socks/streams/forward/client.does.not.connect.no.acceptable.methods/client.rpt:
```diff
@@ -0,0 +1,25 @@
+  option nukleus:window 65536
+  option nukleus:transmission "duplex"
+
+connected
```
I would expect the socks nukleus should inform the client that the underlying socks connection failed with no acceptable methods, either by sending a Reset or an Abort. So in the script I would expect to see an extra line: `read aborted` or `write aborted`. I think best would be reset (write aborted), since socks nukleus will only reply to the initial Begin with a Window frame if it successfully issues the socks connect, so the logical opposite when the socks connect fails would be a Reset.

cmebarrow replied:
done

cmebarrow reviewed (commented):
Please react to my two comments so I can go ahead and approve

Anisotrop committed:
Implement new review feedback.

Anisotrop committed:
Update dependencies versions.

Anisotrop committed:
Change negative test to 'half-duplex'.

jfallows reviewed (approved):
Merging this initial set of spec scripts, pending further iteration before release.
